src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java

```diff
@@ -753,11 +745,12 @@ public BaseResponseBean advanceSessionWithEndpoint(MenuSession menuSession, Stri
         // Sync requests aren't run when executing operations, so stop and check for them after each operation
         for (StackOperation op : endpoint.getStackOperations()) {
             sessionWrapper.executeStackOperations(new Vector<>(Arrays.asList(op)), evalContext);
-            Screen s = menuSession.getNextScreen();
-            if (s instanceof FormplayerSyncScreen) {
+            Screen screen = menuSession.getNextScreen();
+            if (screen instanceof FormplayerSyncScreen) {
                 try {
-                    s.init(sessionWrapper);
-                    doSyncGetNext((FormplayerSyncScreen)s, menuSession);
+                    screen.init(sessionWrapper);
+                    doPostAndSync((FormplayerSyncScreen)screen);
+                    executeAndRebuildSession(menuSession);
```

Comment:
I did think about that but it's not necessary for endpoints. Endpoints are composed of a series of 'push' operations. The last push op is the actual navigation, any ops before that only serve the purpose of performing case claims so there is no session advancement required.

```xml
<stack>
    <push>
        <datum id="parent_id" value="$parent_id"/>
        <command value="'claim_command.my_form.parent_id'"/>
    </push>
    <push>
        <datum id="case_id" value="$case_id"/>
        <command value="'claim_command.my_form.case_id'"/>
    </push>
    <push>
        <command value="'m0'"/>
        <command value="'m1'"/>
        <datum id="parent_id" value="$parent_id"/>
        <datum id="case_id" value="$case_id"/>
        <command value="'m1-f0'"/>
    </push>
</stack>
```

I could put the session advancing logic back if you think it's better or makes it more future proof. The main reason I took it out is because it was doing extra work that's not necessary and it makes the code harder to reason about.

Reply:
that makes sense and no need to revert 👍